src/mesh/mod.rs

```diff
@@ -189,59 +208,79 @@ impl ShapeMesh {
                 &MeshStatus::Ready(ref view) |
                 &MeshStatus::Requested { old_view: Some(ref view) } => {
                     view.draw(surface, camera, env, &self.renderer)?;
+                    // TODO: Debug mode can be removed soon.
                     if self.show_debug {
-                        let highlight = focus_point
-                            .map(|fp| span.contains(fp))
-                            .unwrap_or(false);
+                        let focus_points = self.get_focii(camera, 1);
+                        let highlight = focus_points.len() > 0 && span.contains(focus_points[0]);
                         self.debug_octree.draw(surface, camera, span, highlight)?;
                     }
                 }
                 _ => (),
             }
         }
 
-
         Ok(())
     }
 
-    /// Returns the point on the shape's surface the camera is currently looking at
-    pub fn get_focus(&self, camera: &Camera) -> Option<Point3<f32>> {
+    /// Returns points on the near plane distributed in a grid. These points are given
+    /// in world coordinates. The number of points returned is focus_points^2.
+    pub fn get_focii(&self, camera: &Camera, focus_points: u8) -> Vec<Point3<f32>> {
         const EPSILON: f32 = 0.000_001;
         const MAX_ITERS: u64 = 100;
 
-        let mut pos = camera.position;
-        for _ in 0..MAX_ITERS {
-            let distance = self.shape.min_distance_from(pos);
-            pos += camera.direction() * distance;
-            if distance < EPSILON {
-                break;
-            }
-        }
-        Some(pos)
+        let (top_left, bottom_right) = camera.near_plane_bb();
+        let (frustum_width, frustum_height) = camera.projection.near_plane_dimension();
+        let size_horizontal = frustum_width / focus_points as f32;
+        let size_vertical = frustum_height / focus_points as f32;
+        let center_diff = (bottom_right - top_left) / (2.0 * focus_points as f32);
+
+        let inv_view_trans = camera.inv_view_transform();
+
+        let vec = iter::cube(focus_points as u32)
+            .map(|(x, y, _)| {
+                let center = top_left + Vector3::new(
+                    x as f32 * size_horizontal,
+                    y as f32 * size_vertical,
+                    0.0,
+                ) + center_diff;
+
+                Point3::from_homogeneous(
+                    inv_view_trans * center.to_homogeneous()
+                )
+            })
+            .filter(|p| {
+                let mut pos = camera.position;
+                let dir = (p - camera.position).normalize();
+
+                for _ in 0..MAX_ITERS {
+                    let distance = self.shape.min_distance_from(pos);
+                    pos += dir * distance;
+                    if distance < EPSILON {
+                        return true;
+                    }
+                }
+                false
+            })
+            .collect();
+        vec
```

Comment:
You can even remove the `let vec =` and `; vec`. Then `.collect()` will be the very last thing in the method.